File: packages/control-plane/src/telemetry.ts
//! Telemetry (Section 10.1 + addendum 02a §1.2): ownership is split.
//! The control plane owns show-state fields; the render node owns clock/perf
//! fields. The merge caches the last engine report with a staleness
//! threshold — a stale report means stub values plus `engineConnected: false`.
//! Field shape is always complete: consumers never see a missing field.

import type { EngineTelemetryFrame } from "./protocol.js";
import type { ControlPlaneState, ShowState } from "./state.js";

export interface TelemetryTick {
  // timing
  ts: number;
  // engine-owned (stubbed when stale)
  /** SPEC §10.1 (v0.4): control-plane owned. */
  showState: ShowState;
  masterClockFrame: number;
  droppedFramesTotal: number;
  renderGpuTimeMs: number;
  decodeSessions: number;
  vramUsedMib: number;
  textureCacheUsedMib: number;
  streamBufferMs: number;
  recordSpaceMib: number;
  masterClockDriftMs: number;
  fallbackActive: boolean;
  degradationRung: number;
  // control-plane-owned
  viewItem: string | null;
  previewItem: string | null;
  /** On-air overlays (SPEC §7.10); surfaces `overlay.show`/`hide`. */
  visibleOverlays: string[];
  streamState: string;
  recordState: string;
  automationHold: boolean;
  qualityProfile: string | null;
  // audio (SPEC §8.9, §8.10, new in v0.3.3)
  audioUnderrunsTotal: number;
  audioDriftMs: number;
  busPeakDbfs: Record<string, number>;
  // addendum fields
  engineConnected: boolean;
  deprecationWarnings: Array<{ command: string; resolvedTo: string; stateVersionAtTime: number }>;
  /** ZERO-COPY: which frame path the record tap took — `"zeroCopy"`,
   *  `"cpuReadback"`, or `"none"` before any take has selected one. Required,
   *  not optional: §10.1.1 says a telemetry consumer must never see a missing
   *  field, and that holds for the control plane's tick as much as the
   *  engine's. */
  recordTapPath: string;
  /** Why that path was chosen, so a fallback is distinguishable from a choice.
   *  `"none"` before any take. Required, same reason. */
  recordTapReason: string;
  /** SPEC §10.1 (v0.4.6): the stream transport's own state — `"live"`,
   *  `"reconnecting"`, `"closed"`, or `"none"` before any stream has started.
   *  Engine-owned; `streamState` above is the commanded one and stays `"live"`
   *  through a redial (§9.5), so this is where a redial is visible. Required,
   *  same reason as `recordTapPath`. */
  streamTransportState: string;
}

/** The stub a telemetry field carries before its subsystem has run (§10.1.1).
 *  Mirrors `nbe_protocol::tap_none`. */
export const TAP_NONE = "none";

/** `streamBufferMs` when no measurement exists — no stream session, or no
 *  fresh engine report (SPEC §10.1, ratified v0.4.6). Negative milliseconds
 *  are impossible, so it never collides with a live session's honest `0`,
 *  "the buffer is empty". Mirrors `nbe_protocol::STREAM_BUFFER_NO_SESSION_MS`. */
export const STREAM_BUFFER_NO_SESSION_MS = -1;

/** How long a cached engine report stays authoritative (default 2 s). */
export const ENGINE_TELEMETRY_TTL_MS = 2000;

export interface EngineReport {
  frame: EngineTelemetryFrame;
  receivedAt: number;
}

export function buildTick(
  state: ControlPlaneState,
  engine: WorldTelemetry,
  now: number,
): TelemetryTick {
  const engineFresh = engine.last !== null && now - engine.last.receivedAt <= ENGINE_TELEMETRY_TTL_MS;
  const f = engineFresh ? engine.last!.frame : null;
  return {
    ts: now,
    showState: state.showState,
    masterClockFrame: f?.masterClockFrame ?? 0,
    droppedFramesTotal: f?.droppedFramesTotal ?? 0,
    renderGpuTimeMs: f?.renderGpuTimeMs ?? 0,
    decodeSessions: f?.decodeSessions ?? 0,
    vramUsedMib: f?.vramUsedMib ?? 0,
    textureCacheUsedMib: f?.textureCacheUsedMib ?? 0,
    // §10.1, ratified v0.4.6: buffered ms while a session exists; -1 when no
    // measurement exists — the engine's no-session value forwarded, or no
    // fresh engine report at all. A live session's 0 means "the buffer is
    // empty" and stays distinguishable. ~~"§10.1 law: buffered ms, 0 when
    // nothing is buffered or no report is fresh. Idle vs drained-live is
    // `streamState`'s to say, on this tick."~~ — the repair round's wording
    // (`f8ff895`), struck per §2c; §10.1 never carried that sentence, and
    // v0.4.6 wrote the one it does.
    streamBufferMs: f?.streamBufferMs ?? STREAM_BUFFER_NO_SESSION_MS,
    recordSpaceMib: f?.recordSpaceMib ?? 0,
    masterClockDriftMs: f?.masterClockDriftMs ?? 0,
    fallbackActive: f?.fallbackActive ?? state.fallbackActive,
    degradationRung: f?.degradationRung ?? 0,
    // ZERO-COPY: always forwarded, stubbed when the engine has not reported
    // one — no stale engine report, or an engine build older than the stub.
    // Still no `?? "cpuReadback"`: `TAP_NONE` is not a path, so a machine that
    // never recorded stays distinguishable from one that fell back to the §0.1
    // assumption 24 allowance. Parsing the field at the boundary is not the
    // same as an operator seeing it, which is what these two lines are for.
    recordTapPath: f?.recordTapPath ?? TAP_NONE,
    recordTapReason: f?.recordTapReason ?? TAP_NONE,
    // SPEC §10.1, v0.4.6. Always forwarded, stubbed — the rule followed is
    // §10.1.1's completeness (the `recordTapPath` FINAL shape, Phase 3b), not
    // the F1-era one it replaced: PR #24's F1 fix forwarded the tap fields
    // only when present, on the reasoning that absence meant "no take yet".
    // §10.1.1 forbids that reasoning, so this is decided against it. A stale
    // report stubs to "none" rather than "closed": with no fresh engine
    // report the control plane does not know what the socket is doing, and
    // `engineConnected: false` already says why.
    streamTransportState: f?.streamTransportState ?? TAP_NONE,
    viewItem: state.viewItem,
    previewItem: state.previewItem,
    visibleOverlays: Array.from(state.visibleOverlays),
    streamState: state.streamState,
    recordState: state.recordState,
    automationHold: state.automationHold,
    // SPEC §10.1.1: the engine's probed (effective) profile wins while its
    // report is fresh; otherwise the manifest's requested profile is the only
    // honest answer the control plane has.
    qualityProfile: f?.qualityProfile ?? state.qualityProfile,
    audioUnderrunsTotal: f?.audioUnderrunsTotal ?? 0,
    audioDriftMs: f?.audioDriftMs ?? 0,
    busPeakDbfs: f?.busPeakDbfs ?? {},
    engineConnected: engineFresh,
    // Deprecation warnings are per-subscriber: the server fans each accepted
    // deprecated command into every subscriber's own cursor and fills this in.
    // A shared drain here would hand the warning to whichever tick fired first.
    deprecationWarnings: [],
  };
}

/** Rolling per-connection telemetry state: the last engine report. */
export interface WorldTelemetry {
  last: EngineReport | null;
}

export function newWorldTelemetry(): WorldTelemetry {
  return { last: null };
}

export function ingestEngineFrame(world: WorldTelemetry, frame: EngineTelemetryFrame, now: number): void {
  world.last = { frame, receivedAt: now };
}

File: crates/nbe-engine/src/telemetry.rs
//! Telemetry (Prompt 03 Step 5, SPEC §10.1.1 [ownership]): the engine is
//! authoritative for clock/perf fields only; the show-state fields are the
//! control plane's. Emit an `engineTelemetry` frame at the cadence Step 5
//! sets.

use crate::state::EngineState;
use nbe_protocol::{EngineFrame, EngineTelemetry};
use std::path::Path;

/// Free space on the record target's volume, in MiB (SPEC §10.1).
///
/// Computed on the telemetry tick (or lazily by the caller holding the record
/// directory) — never in the frame path, so the render loop keeps its
/// deadline behavior unchanged. Unwritable/missing targets degrade to `0.0`
/// here; the loud `E_DISK` refusal lives in
/// [`crate::record::available_space_mib`], which this wraps.
pub fn record_space_mib_for(dir: Option<&Path>) -> f64 {
    dir.and_then(|d| crate::record::available_space_mib(d).ok())
        .unwrap_or(0.0)
}

/// Build one engine-owned telemetry frame, measuring `recordSpaceMib`
/// against the real target volume when `record_dir` is `Some`.
pub fn build_tick_for_dir(state: &EngineState, record_dir: Option<&Path>) -> EngineFrame {
    // Read the tap selection ONCE, before the struct literal. Locking twice
    // inside it deadlocks: struct-literal temporaries live until the end of the
    // enclosing statement, so the first `MutexGuard` is still held when the
    // second `lock()` runs, on a non-reentrant `Mutex`. Found by
    // `pump_tick_wires_the_loaded_record_dir` hanging rather than failing.
    let tap = *state.record_tap_selection.lock().unwrap();
    // streamBufferMs (§10.1, law): the live session's admitted-but-unwritten
    // bytes through the stream's envelope bitrate (channel backlog INCLUDED —
    // see the rtmp module docs). With no session the buffer holds nothing,
    // and 0.0 says exactly that — a measurement of an absent buffer, not a
    // stub standing in for one. Whether the stream is idle or drained-live is
    // `streamState`'s job (the control plane's, "as commanded"), already on
    // the same tick.
    //
    // PR #30 first shipped -1.0 here as a "NO-SESSION sentinel". That changed
    // the meaning of a ratified field inside a feature PR, which is the
    // user's change to make, not ours; it is reverted and drafted as an
    // UNRATIFIED candidate in `docs/v0.5-outline.md` §7 instead.
    //
    // Read under one short lock; the counter itself is atomic, so the tick
    // never waits on the socket. `streamTransportState` (§10.1, v0.4.6) is
    // read in the same scope: the publisher's own view of the socket, which
    // `streamState` ("as commanded") deliberately does not follow (§9.5).
    let (stream_buffer_ms, transport) = {
        let session = state.stream_session.lock().unwrap();
        (
            session
                .as_ref()
                .map(|s| s.stream_buffer_ms())
                .unwrap_or(0.0),
            session.as_ref().map(|s| s.publisher_state()),
        )
    };
    // No session: `"closed"` once any stream has opened (the socket is gone),
    // the `"none"` stub before one ever has — §10.1.1, the `recordTapPath`
    // precedent. `"none"` is not a transport state, so never-streamed stays
    // distinguishable from closed.
    let stream_transport_state = match transport {
        Some(p) => p.as_str().to_string(),
        None if state
            .stream_transport_opened
            .load(std::sync::atomic::Ordering::SeqCst) =>
        {
            crate::record::rtmp::PublisherState::Closed
                .as_str()
                .to_string()
        }
        None => nbe_protocol::tap_none(),
    };
    let frame = EngineTelemetry {
        master_clock_frame: state.master_frame().unwrap_or(0),
        dropped_frames_total: state
            .dropped_frames_total
            .load(std::sync::atomic::Ordering::SeqCst),
        render_gpu_time_ms: 0.0,
        decode_sessions: state.sessions.active(),
        vram_used_mib: 0.0,
        texture_cache_used_mib: 0.0,
        stream_buffer_ms,
        record_space_mib: record_space_mib_for(record_dir),
        master_clock_drift_ms: 0.0,
        fallback_active: state
            .fallback_active
            .load(std::sync::atomic::Ordering::SeqCst),
        degradation_rung: state.degradation_rung(),
        // Effective probe result from GPU init, capped by the manifest's
        // requested profile. The engine is the engine — it is authoritative.
        quality_profile: *state.quality_profile.lock().unwrap(),
        // SPEC §8.10 / §8.9 / §10.1: measured by the audio graph, published
        // here. Zero means "no audio engine running", not "no problem".
        audio_underruns_total: state
            .audio_underruns_total
            .load(std::sync::atomic::Ordering::SeqCst),
        audio_drift_ms: f64::from_bits(
            state
                .audio_drift_ms_bits
                .load(std::sync::atomic::Ordering::SeqCst),
        ),
        // A snapshot: the driver closes each meter window on its own boundary,
        // so reading here neither ends a window nor races another reader.
        bus_peak_dbfs: state.bus_peaks.lock().unwrap().clone(),
        // ZERO-COPY Phase 2: which path the record tap took, and why. Absent
        // until a take selects one — an operator reading this can tell a
        // machine that chose zero-copy from one that fell back to the §0.1
        // assumption 24 allowance, which is the whole reason it is reported.
        // §10.1.1: always emitted. `"none"` before a take selects — a stub, not
        // an omission, because an absent field and a stubbed field are
        // different failures and only one of them is diagnosable.
        record_tap_path: tap
            .map(|s| s.path.as_str().to_string())
            .unwrap_or_else(nbe_protocol::tap_none),
        record_tap_reason: tap
            .map(|s| format!("{:?}", s.reason))
            .unwrap_or_else(nbe_protocol::tap_none),
        stream_transport_state,
    };
    EngineFrame::EngineTelemetry {
        v: nbe_protocol::PROTOCOL_VERSION.to_string(),
        ts: std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_millis() as f64)
            .unwrap_or(0.0),
        fields: frame,
    }
}

/// Build one engine-owned telemetry frame from the current state. The shape
/// must be complete even when values are stubs — a consumer should never see
/// a missing field (§5.9.1 + §10.1.1).
///
/// Render-loop path: unchanged. No record target is known here, so
/// `recordSpaceMib` reports `0.0`; callers holding the record directory use
/// [`build_tick_for_dir`]. The channel pump reads the loaded package's record
/// directory out of engine state and calls [`build_tick_for_dir`] with it.
pub fn build_tick(state: &EngineState) -> EngineFrame {
    build_tick_for_dir(state, None)
}
